Sprint-3/4-stretch-investigate/password-validator.test.js

```diff
@@ -15,12 +15,59 @@ To be valid, a password must:
 You must breakdown this problem in order to solve it. Find one test case first and get that working
 */
 const isValidPassword = require("./password-validator");
-test("password has at least 5 characters", () => {
-    // Arrange
-    const password = "12345";
-    // Act
-    const result = isValidPassword(password);
-    // Assert
-    expect(result).toEqual(true);
-}
-);
\ No newline at end of file
+
+// test("password has at least 5 characters", () => {
+//     // Arrange
+//     const password = "12345";
+//     // Act
+//    const result = isValidPassword(password);
+//     // Assert
+//     expect(result).toBe(true);
+// }
+// );
+test("password has a unknown character", () => {
+  const password = "Mill3@";
+  const result = isValidPassword(password);
+  expect(result).toBe(false);
+
+})
+
+
+ test("password has at least 5 characters", () => {
+  expect(isValidPassword("12345")).toBe(false);// Fails because no letters or special chars
+  expect(isValidPassword("1234")).toBe(false); // Fails because the password is less than 5 characters.
+});
+
+test("password has at least one uppercase letter", () => {
+  expect(isValidPassword("password1!")).toBe(false); // No uppercase letter
+});
+
+test("password has at least one lowercase letter", () => {
+  expect(isValidPassword("PASSWORD1!")).toBe(false); // No lowercase letter
+});
+
+test("password has at least one number", () => {
+  expect(isValidPassword("Password!")).toBe(false); // No number
+});
+
+test("password has at least one special character", () => {
+  expect(isValidPassword("Password1")).toBe(false); // No special character
+});
+
+
+test("valid password", () => {
+  expect(isValidPassword("Valid1!")).toBe(true); // Meets all conditions
+});
+
+test("password should not be a previously used password", () => {
+  expect(isValidPassword("Valid1!")).toBe(false); // Used before
```

Comment:
This works due to the global state (bannedPassword array), but it can make your tests dependent on each other, which is not ideal for unit testing.
💭 Consider passing an array of banned passwords as a function argument instead. This would make the functino more robust (for this exercise, at least).